changes_test.go

```diff
@@ -3265,6 +3265,261 @@ func (s *changesSuite) TestFromJujuMassiveUnitColocation(c *gc.C) {
+        machines:
+            0:
+            1:
+            `
+	existingModel := &bundlechanges.Model{
+		Applications: map[string]*bundlechanges.Application{
+			"memcached": {
+				Name:  "memcached",
+				Charm: "cs:xenial/mem-47",
+				Units: []bundlechanges.Unit{
+					{Name: "memcached/1", Machine: "1"},
+					{Name: "memcached/2", Machine: "2"},
+					{Name: "memcached/3", Machine: "3"},
+				},
+			},
+		},
+		Machines: map[string]*bundlechanges.Machine{
+			"1": {ID: "1"},
+			"2": {ID: "2"},
+			"3": {ID: "3"},
+		},
+		MachineMap: map[string]string{
+			// using --map-machines=existing
+			"1": "1",
+			"2": "2",
+			"3": "3",
+			// Plus an explicit mapping.
+			"0": "3",
+		},
+	}
+	// Now that we have a consistent mapping, no changes are needed.
+	s.checkBundleExistingModel(c, bundleContent, existingModel, nil)
+}
+
+func (s *changesSuite) TestContainerHosts(c *gc.C) {
+	// If we have a bundle that needs to create a container, we don't
+	// treat the machine hosting the container as not having
+	// dependants.
+	bundleContent := `
+        applications:
+            memcached:
+                charm: cs:xenial/mem-47
+                num_units: 2
+                to: [1, "lxd:2"]
+        machines:
+            1:
+            2:
+            `
+	existingModel := &bundlechanges.Model{
+		Applications: map[string]*bundlechanges.Application{
+			"memcached": {
+				Name:  "memcached",
+				Charm: "cs:xenial/mem-47",
+				Units: []bundlechanges.Unit{
+					{Name: "memcached/1", Machine: "1"},
+				},
+			},
+		},
+		Machines: map[string]*bundlechanges.Machine{
+			"1": {ID: "1"},
+		},
+		MachineMap: map[string]string{
+			// using --map-machines=existing
+			"1": "1",
+		},
+	}
+	expectedChanges := []string{
+		"add new machine 2",
+		"add lxd container 2/lxd/0 on new machine 2",
+		"add unit memcached/2 to 2/lxd/0",
+	}
+	s.checkBundleExistingModel(c, bundleContent, existingModel, expectedChanges)
+}
+
+func (s *changesSuite) TestSingleTarget(c *gc.C) {
+	bundleContent := `
+        applications:
+            memcached:
+                charm: cs:xenial/mem-47
+                num_units: 2
+                to: [0, 1]
+        machines:
+            0:
+            1:
+            `
+	existingModel := &bundlechanges.Model{
+		Applications: map[string]*bundlechanges.Application{
+			"memcached": {
+				Name:  "memcached",
+				Charm: "cs:xenial/mem-47",
+				Units: []bundlechanges.Unit{
+					{Name: "memcached/1", Machine: "1"},
+					{Name: "memcached/2", Machine: "2"},
+				},
+			},
+		},
+		Machines: map[string]*bundlechanges.Machine{
+			"1": {ID: "1"},
+			"2": {ID: "2"},
+		},
+		MachineMap: map[string]string{
+			// using --map-machines=existing
+			"1": "1",
+			"2": "2",
+		},
+	}
+	s.checkBundleImpl(c, bundleContent, existingModel, nil, `bundle and machine mapping are inconsistent: need an explicit entry mapping bundle machine "0" to unreferenced model machine "2" - the target should host \[memcached\]`)
```

Comment:
I do wonder whether people will complain that we don't just do that. I'd prefer if we indicated our uncertainty in the error message.

Perhaps adding ", perhaps" after the bundle machine, so it would read:

    bundle and machine mapping are inconsistent: need an explicit entry mapping bundle machine "0", perhaps to unreferenced model machine "2" - the target should host \[memcached\]`

I also wonder about the hypen... but I guess it is ok

Reply:
Yeah, I thought about that too. That wording's better.